pkg/integrations/prometheus/promql/parser.go

```diff
@@ -0,0 +1,564 @@
+	}
+
+	// Handle grouping
+	for _, grp := range expr.Grouping {
+		tagFilter := structs.TagsFilter{
+			TagKey:          grp,
+			RawTagValue:     "*",
+			HashTagValue:    xxhash.Sum64String("*"),
+			TagOperator:     segutils.TagOperator(segutils.Equal),
+			LogicalOperator: segutils.And,
+		}
+		mQuery.TagsFilters = append(mQuery.TagsFilters, &tagFilter)
+	}
+	if len(expr.Grouping) > 0 {
+		mQuery.Groupby = true
+	}
+
+	mQueryAgg := &structs.MetricQueryAgg{
+		AggBlockType:    structs.AggregatorBlock,
+		AggregatorBlock: mQuery.Aggregator.Clone(),
+	}
+
+	// If MQueryAggs is nil, set it to the new MetricQueryAgg
+	if mQuery.MQueryAggs == nil {
+		mQuery.MQueryAggs = mQueryAgg
+	} else {
+		// Otherwise, set the new MetricQueryAgg to the head of the chain
+		// and set the current head as the next of the new MetricQueryAgg
+		mQueryAgg.Next = mQuery.MQueryAggs
+		mQuery.MQueryAggs = mQueryAgg
+	}
+
+	// Reset the Aggregator to handle the next aggregation correctly
+	mQuery.Aggregator = structs.Aggregation{}
+}
+
+func handleCallExpr(call *parser.Call, mQuery *structs.MetricsQuery) error {
+	var err error
+	defaultCase := false
+
+	for _, arg := range call.Args {
+		switch arg := arg.(type) {
+		case *parser.MatrixSelector:
+			err = handleCallExprMatrixSelectorNode(call, mQuery)
+		case *parser.VectorSelector:
+			err = handleCallExprVectorSelectorNode(call, mQuery)
+		case *parser.ParenExpr:
+			err = handleCallExprParenExprNode(call, arg, mQuery)
+		case *parser.SubqueryExpr:
+			err = handleCallExprMatrixSelectorNode(call, mQuery)
+		default:
+			defaultCase = true
+		}
+		if err != nil {
+			return fmt.Errorf("handleCallExpr: cannot parse Call Node: %v", err)
+		}
+
+		// Break if the case is not default. As Call Expr can have multiple arguments
+		// And they will be procesed based on the Node type.
+		if !defaultCase {
+			break
+		}
+	}
+
+	if defaultCase {
+		// If the default case is true, then args are not of type MatrixSelector or VectorSelector
+		// It might be AggregateExpr. And the function can be a Range or Math Function.
+		// So, we need to check for both the cases.
+		err = handleCallExprVectorSelectorNode(call, mQuery)
+		if err != nil {
+			err = handleCallExprMatrixSelectorNode(call, mQuery)
+		}
+	}
+
+	if err != nil {
+		return fmt.Errorf("handleCallExpr: cannot parse Call Node: %v", err)
+	}
+
+	mQueryAgg := &structs.MetricQueryAgg{
+		AggBlockType:  structs.FunctionBlock,
+		FunctionBlock: mQuery.Function.Clone(),
+	}
+
+	// If MQueryAggs is nil, set it to the new MetricQueryAgg
+	if mQuery.MQueryAggs == nil {
+		mQuery.MQueryAggs = mQueryAgg
+	} else {
+		// Otherwise, set the new MetricQueryAgg to the head of the chain
+		// and set the current head as the next of the new MetricQueryAgg
+		mQueryAgg.Next = mQuery.MQueryAggs
+		mQuery.MQueryAggs = mQueryAgg
+	}
+
+	// Reset the Function to handle the next function call correctly
+	mQuery.Function = structs.Function{}
+
+	return nil
```

Comment:
Why is this function (`handleCallExpr()`) responsible for updating the `mQuery` parameter like this? I'd recommend moving this logic to a new function. Then probably refactor so you call that new function outside of `handleCallExpr()`, but if you need to call it within `handleCallExpr()` you'll probably want to add a comment to the definition of `handleCallExpr()` saying how it updates `mQuery`, or return the new `mQuery` instead of just updating the one sent as a parameter

Reply:
I will move this logic to a new function. As I am doing something similar in the other function `handleAggregateExpr`.